File: spynnaker_visualisers/heat/menu.cpp
enum menuentries_t {
    MENU_SEPARATOR,
    // File Submenu Items
    FILE_SAVE_SPINN,
    FILE_SAVE_NEURO,
    FILE_RESUME,
    FILE_PAUSE,
    FILE_END,
    // Transform Submenu Items
    XFORM_XFLIP,
    XFORM_YFLIP,
    XFORM_VECTORFLIP,
    XFORM_ROTATEFLIP,
    XFORM_REVERT,
    // Main Menu Items
    MENU_BORDER_TOGGLE,
    MENU_NUMBER_TOGGLE,
    MENU_FULLSCREEN_TOGGLE,
    MENU_PAUSE,
    MENU_RESUME,
    MENU_QUIT
};

static void menu_callback(int value)
{
    switch (value) {
    case XFORM_XFLIP:
	xflip = !xflip;
	break;
    case XFORM_YFLIP:
	yflip = !yflip;
	break;
    case XFORM_VECTORFLIP:
	vectorflip = !vectorflip;
	break;
    case XFORM_ROTATEFLIP:
	rotateflip = !rotateflip;
	break;
    case XFORM_REVERT:
	cleardown();
	break;
    case MENU_BORDER_TOGGLE:
	gridlines = !gridlines;
	break;
    case MENU_NUMBER_TOGGLE:
	plotvaluesinblocks = !plotvaluesinblocks;
	break;
    case MENU_FULLSCREEN_TOGGLE:
	if (fullscreen) {
	    windowBorder = oldwindowBorder;	// restore old bordersize
	    windowWidth -= keyWidth;		// recover the key area
	    plotWidth = windowWidth - 2 * windowBorder - keyWidth;
	} else {
	    oldwindowBorder = windowBorder;	// used as border disappears when going full-screen
	    windowBorder = 0;			// no borders around the plots
	    windowWidth += keyWidth;		// take over the area used for the key too
	    plotWidth = windowWidth - keyWidth;
	}
	fullscreen = !fullscreen;
	break;
    case MENU_PAUSE:
	// send pause packet out
	for (int i = 0 ; i < all_desired_chips() ; i++) {
	    send_to_chip(i, 0x21, 2, 0, 0, 0, 4, 0, 0, 0, 0);
	}
	freezedisplay = 1;
	freezetime = timestamp();		// get time now in us
	break;
    case MENU_RESUME:
	// send resume/restart packet out
	for (int i = 0 ; i < all_desired_chips() ; i++) {
	    send_to_chip(i, 0x21, 3, 0, 0, 0, 4, 0, 0, 0, 0);
	}
	freezedisplay = 0;
	break;
    case MENU_QUIT:
	safelyshut();
	break;
    }

    needtorebuildmenu = 1;
}

static void rebuildmenu(void)
{
    glutDestroyMenu(RHMouseMenu);
    RHMouseMenu = glutCreateMenu(menu_callback);

    glutAddMenuEntry("(X) Mirror (left to right swap)", XFORM_XFLIP);
    glutAddMenuEntry("(Y) Reflect (top to bottom swap)", XFORM_YFLIP);
    glutAddMenuEntry("(V) Vector Swap (Full X+Y Reversal)", XFORM_VECTORFLIP);
    glutAddMenuEntry("90 (D)egree Rotate Toggle", XFORM_ROTATEFLIP);
    glutAddMenuEntry("(C) Revert changes back to default", XFORM_REVERT);
    glutAddMenuEntry("-----", MENU_SEPARATOR);
    if (gridlines) {
	glutAddMenuEntry("Grid (B)orders off", MENU_BORDER_TOGGLE);
    } else {
	glutAddMenuEntry("Grid (B)orders on", MENU_BORDER_TOGGLE);
    }
    if (plotvaluesinblocks) {
	glutAddMenuEntry("Numbers (#) off", MENU_NUMBER_TOGGLE);
    } else {
	glutAddMenuEntry("Numbers (#) on", MENU_NUMBER_TOGGLE);
    }
    if (fullscreen) {
	glutAddMenuEntry("(F)ull Screen off", MENU_FULLSCREEN_TOGGLE);
    } else {
	glutAddMenuEntry("(F)ull Screen on", MENU_FULLSCREEN_TOGGLE);
    }

    glutAddMenuEntry("-----", MENU_SEPARATOR);

    if (!freezedisplay) {
	glutAddMenuEntry("(\") Pause Plot", MENU_PAUSE);
    } else {
	glutAddMenuEntry("(P)lay / Restart Plot", MENU_RESUME);
    }
    glutAddMenuEntry("(Q)uit", MENU_QUIT);
    glutAttachMenu (GLUT_RIGHT_BUTTON);
}

static void logifmenuopen(int status, int x, int y)
{
    use(x);
    use(y);

    menuopen = int(status == GLUT_MENU_IN_USE);
    if (!menuopen && needtorebuildmenu) {
	// if menu is not open we can make changes
	rebuildmenu();
	needtorebuildmenu = 0;
    }
}

File: spynnaker_visualisers/heat/menu.py
from OpenGL.GLUT import glutCreateMenu, glutDestroyMenu, glutAddMenuEntry, \
    glutAttachMenu, GLUT_RIGHT_BUTTON, GLUT_MENU_IN_USE
import spynnaker_visualisers.heat.state as state
import spynnaker_visualisers.heat.sdp as sdp 
from spynnaker_visualisers.heat import utils
from spynnaker_visualisers.heat.visualiser import safelyshut
from spynnaker_visualisers.heat.events import toggle_fullscreen


_RHMouseMenu = None
_needtorebuildmenu = False
_menuopen = False
# Should be enum...
MENU_SEPARATOR = 0
XFORM_XFLIP = 1
XFORM_YFLIP = 2
XFORM_VECTORFLIP = 3
XFORM_ROTATEFLIP = 4
XFORM_REVERT = 5
MENU_BORDER_TOGGLE = 6
MENU_NUMBER_TOGGLE = 7
MENU_FULLSCREEN_TOGGLE = 8
MENU_PAUSE = 9
MENU_RESUME = 10
MENU_QUIT = 11


def callback(option):
    if option == XFORM_XFLIP:
        state.xflip = not state.xflip
    elif option == XFORM_YFLIP:
        state.yflip = not state.yflip
    elif option == XFORM_VECTORFLIP:
        state.vectorflip = not state.vectorflip
    elif option == XFORM_ROTATEFLIP:
        state.rotateflip = not state.rotateflip
    elif option == XFORM_REVERT:
        state.cleardown()
    elif option == MENU_BORDER_TOGGLE:
        state.gridlines = not state.gridlines
    elif option == MENU_NUMBER_TOGGLE:
        state.plotvaluesinblocks = not state.plotvaluesinblocks
    elif option == MENU_FULLSCREEN_TOGGLE:
        toggle_fullscreen()
    elif option == MENU_PAUSE:
        for i in sdp.all_desired_chips():
            sdp.send_to_chip(i, 0x21, 2, 0, 0, 0, 0, 0, 0, 0)
        state.freezedisplay = True
        state.freezetime = utils.timestamp()
    elif option == MENU_RESUME:
        for i in sdp.all_desired_chips():
            sdp.send_to_chip(i, 0x21, 3, 0, 0, 0, 0, 0, 0, 0)
        state.freezedisplay = False
    elif option == MENU_QUIT:
        safelyshut()
    _needtorebuildmenu = True


def rebuild():
    global _RHMouseMenu
    glutDestroyMenu(_RHMouseMenu)
    _RHMouseMenu = glutCreateMenu(callback)

    glutAddMenuEntry("(X) Mirror (left to right swap)", XFORM_XFLIP)
    glutAddMenuEntry("(Y) Reflect (top to bottom swap)", XFORM_YFLIP)
    glutAddMenuEntry("(V) Vector Swap (Full X+Y Reversal)", XFORM_VECTORFLIP)
    glutAddMenuEntry("90 (D)egree Rotate Toggle", XFORM_ROTATEFLIP)
    glutAddMenuEntry("(C) Revert changes back to default", XFORM_REVERT)
    glutAddMenuEntry("-----", MENU_SEPARATOR)
    if state.gridlines:
        glutAddMenuEntry("Grid (B)orders off", MENU_BORDER_TOGGLE)
    else:
        glutAddMenuEntry("Grid (B)orders on", MENU_BORDER_TOGGLE)
    if state.plotvaluesinblocks:
        glutAddMenuEntry("Numbers (#) off", MENU_NUMBER_TOGGLE)
    else:
        glutAddMenuEntry("Numbers (#) on", MENU_NUMBER_TOGGLE)
    if state.fullscreen:
        glutAddMenuEntry("(F)ull Screen off", MENU_FULLSCREEN_TOGGLE)
    else:
        glutAddMenuEntry("(F)ull Screen on", MENU_FULLSCREEN_TOGGLE)
    glutAddMenuEntry("-----", MENU_SEPARATOR)
    if not state.freezedisplay:
        glutAddMenuEntry("(\") Pause Plot", MENU_PAUSE)
    else:
        glutAddMenuEntry("(P)lay / Restart Plot", MENU_RESUME)
    glutAddMenuEntry("(Q)uit", MENU_QUIT)
    glutAttachMenu(GLUT_RIGHT_BUTTON)
    return _RHMouseMenu


def logifmenuopen(status, x, y):  # @UnusedVariable
    global _menuopen
    _menuopen = (status == GLUT_MENU_IN_USE)
    rebuild_if_needed()

def trigger_rebuild():
    _needtorebuildmenu = True

def rebuild_if_needed():
    if _needtorebuildmenu and not _menuopen:
        rebuild()
        _needtorebuildmenu = False
